cpp/src/indexing/point_quadtree.cu

```diff
@@ -0,0 +1,803 @@
+ * extra care on minmizing peak device memory usage by deallocating memory as
+ * early as possible
+ */
+
+namespace {
+
+template <typename Vector>
+inline auto shrink_vector(Vector &v, cudf::size_type size) {
+  v.resize(size, 0);
+  v.shrink_to_fit();
+  return v;
+}
+
+template <typename... Ts>
+inline auto make_zip_iterator(Ts... its) {
+  return thrust::make_zip_iterator(
+      thrust::make_tuple(std::forward<Ts>(its)...));
+}
+
+template <typename T>
+inline std::unique_ptr<cudf::column> make_fixed_width_column(
+    cudf::size_type size, cudaStream_t stream = 0,
+    rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource()) {
+  return cudf::make_fixed_width_column(
+      cudf::data_type{cudf::experimental::type_to_id<T>()}, size,
+      cudf::mask_state::UNALLOCATED, stream, mr);
+}
+
+template <typename T>
+inline rmm::device_vector<uint32_t> compute_point_keys(
+    cudf::mutable_column_view &x, cudf::mutable_column_view &y, T const x1,
+    T const y1, T const x2, T const y2, T const scale,
+    cudf::size_type const num_levels, cudaStream_t stream) {
+  // Compute Morton codes (z-order) for each point
+  auto policy = rmm::exec_policy(stream);
+  rmm::device_vector<uint32_t> keys(x.size());
+  auto points = make_zip_iterator(x.begin<T>(), y.begin<T>());
+  thrust::transform(
+      policy->on(stream), points, points + x.size(), keys.begin(),
+      [x1, y1, x2, y2, num_levels, scale] __device__(auto const &point) {
+        T x, y;
+        thrust::tie(x, y) = point;
+        if (x < x1 || x > x2 || y < y1 || y > y2) {
+          // If the point is outside the bbox, return a max_level key
+          return static_cast<uint32_t>((1 << (2 * num_levels)) - 1);
+        }
+        return z_order((x - x1) / scale, (y - y1) / scale);
+      });
+
+  // Sort the points and codes
+  thrust::sort_by_key(policy->on(stream), keys.begin(), keys.end(), points);
+
+  return keys;
+}
+
+template <typename InputIterator1, typename InputIterator2,
+          typename OutputIterator1, typename OutputIterator2,
+          typename BinaryPred>
+inline cudf::size_type construct_quadrant(
+    InputIterator1 keys_begin, InputIterator1 keys_end, InputIterator2 vals_in,
+    OutputIterator1 keys_out, OutputIterator2 vals_out, BinaryPred binary_op,
+    cudaStream_t stream) {
+  auto policy = rmm::exec_policy(stream);
+  auto result = thrust::reduce_by_key(policy->on(stream), keys_begin, keys_end,
+                                      vals_in, keys_out, vals_out,
+                                      thrust::equal_to<uint32_t>(), binary_op);
+  return thrust::distance(keys_out, result.first);
+}
+
+template <typename T>
+struct tuple_sum {
+  inline __device__ thrust::tuple<T, T> operator()(
+      thrust::tuple<T, T> const &a, thrust::tuple<T, T> const &b) {
+    return thrust::make_tuple(thrust::get<0>(a) + thrust::get<0>(b),
+                              thrust::get<1>(a) + thrust::get<1>(b));
+  }
+};
```

Comment:
Can this just be an `operator+`?

```suggestion
template <typename T>
__device__ thrust::tuple<T, T> operator+(
      thrust::tuple<T, T> const &a, thrust::tuple<T, T> const &b) {
    return thrust::make_tuple(thrust::get<0>(a) + thrust::get<0>(b),
                              thrust::get<1>(a) + thrust::get<1>(b));
  }
};
```

Reply:
I guess this would still require instantiating a thrust::plus of appropriate type:

```
template <typename T>
using tuple_sum<T> = thrust::plus<thrust:tuple<T, T>>;
```

Or you could pass the thing on the right of the = to whereever you currently pass `tuple_sum<T>`

Doesn't buy or save much.